modules/navigation/nav_map.cpp

```diff
@@ -61,15 +62,23 @@ void NavMap::set_cell_size(float p_cell_size) {
 	regenerate_polygons = true;
 }
 
+void NavMap::set_cell_height(float p_cell_height) {
+	cell_height = p_cell_height;
+	regenerate_polygons = true;
+}
+
 void NavMap::set_edge_connection_margin(float p_edge_connection_margin) {
 	edge_connection_margin = p_edge_connection_margin;
 	regenerate_links = true;
 }
 
 gd::PointKey NavMap::get_point_key(const Vector3 &p_pos) const {
-	const int x = int(Math::floor(p_pos.x / cell_size));
-	const int y = int(Math::floor(p_pos.y / cell_size));
-	const int z = int(Math::floor(p_pos.z / cell_size));
+	const real_t raw_cell_key_x = p_pos.x / cell_size;
+	const real_t raw_cell_key_y = p_pos.y / cell_height;
+	const real_t raw_cell_key_z = p_pos.z / cell_size;
+	const int x = int(Math::is_equal_approx(raw_cell_key_x, Math::ceil(raw_cell_key_x)) ? Math::ceil(raw_cell_key_x) : Math::floor(raw_cell_key_x));
+	const int y = int(Math::is_equal_approx(raw_cell_key_y, Math::ceil(raw_cell_key_y)) ? Math::ceil(raw_cell_key_y) : Math::floor(raw_cell_key_y));
+	const int z = int(Math::is_equal_approx(raw_cell_key_z, Math::ceil(raw_cell_key_z)) ? Math::ceil(raw_cell_key_z) : Math::floor(raw_cell_key_z));
```

Comment:
With this approach it uses the `Math::ceil(raw_cell_key_...)` value only if the `raw_cell_key_...` is near to that value, which make this algorithm not that different from the original one: for example if the value is `0.9` it will still use `0` (floor) as value, since `0.9` is not enough near to `1.0` (ceil).

I wonder if the following one is better, considering that it returns always the closest integer:
```c++
	const int x = static_cast<int>(Math::round(p_pos.x / cell_size));
	const int y = static_cast<int>(Math::round(p_pos.y / cell_size));
	const int z = static_cast<int>(Math::round(p_pos.z / cell_size));
```

If the value is `0.9` it will use `1.0` as round returns the nearest integral. What do you think?

Reply:
Well, haven't thought about it but now that I think of it, it actually should work and the code will be much cleaner. I'll give it a try.

Reply:
Thinking more about it I just realized that my above proposed solution will likely change how the algorithm is supposed to work, and that will cause some issues. In any case I still think we need a much simpler way to obtain that `PointKey`. The following one is a similar solution to the one you proposed, what do you think? 
```cpp
const int x = static_cast<int>((p_pos.x / cell_size) + CMP_EPSILON);
const int y = static_cast<int>((p_pos.y / cell_size) + CMP_EPSILON);
const int z = static_cast<int>((p_pos.z / cell_size) + CMP_EPSILON);
```

Reply:
@AndreaCatania I've tested the solution with `Math::round` and it's both - fine (tested on various cases) and clean (easy to read and understand + it properly handles precision) and thus I think that will do.